### ISSUE TITLE
Annotations fail to anchor on chrome

### ISSUE THREAD

**Issue by dwhly** [hypothesis/h#2129]
On chrome, go here:
https://via.hypothes.is/h/http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/

No annotations.  Even thought the stream shows some here:
https://hypothes.is/stream?q=uri:http:%2F%2Farstechnica.com%2Fgaming%2F2015%2F04%2Fvalve-software-co-founder-gabe-newell-becomes-a-video-game-character%2F

On safari, the annotations show.  Tried OSX / Safari 8 via Saucelabs.

**Comment by RawKStar77** [comment]
A possibly important detail: these annotations were made with Safari.

**Comment by nickstenning** [comment]
I don't see annotations at either link.

**Comment by dwhly** [comment]
@RawKStar77 Can you and I try to repro this again?

**Comment by judell** [comment]
See also: 

http://www.a2gov.org/services/Pages/OpenCityHall.aspx?pd_url=http%3a//www.peakdemocracy.com/portals/116/forum_home?admin_mode%3Doff#peak_democracy

Compare:

1. What the extension sees

2. https://hypothes.is/stream?q=uri:http:%2F%2Fwww.a2gov.org%2Fservices%2FPages%2FOpenCityHall.aspx%3Fpd_url%3Dhttp%253a%2F%2Fwww.peakdemocracy.com%2Fportals%2F116%2Fforum_home%3Fadmin_mode%253Doff%23peak_democracy

3. https://hypothes.is/api/search?uri=http%3A%2F%2Fwww.a2gov.org%2Fservices%2FPages%2FOpenCityHall.aspx%3Fpd_url%3Dhttp%253a%2F%2Fwww.peakdemocracy.com%2Fportals%2F116%2Fforum_home%3Fadmin_mode%253Doff%23peak_democracy

4. # of annotations displayed in bucket (1) vs seen in extension (3)

**Comment by gergely-ujvari** [comment]
We should separate the two phenomena, as they're separate issues, namely the failed anchoring and the problems with our search api. (Stream uses our search api for loading the initial annotations so there should be no difference between those two).

**Comment by gergely-ujvari** [comment]
I checked @judell's link to a2gov and found out the following:

If I check the [a2gov page](http://www.a2gov.org/services/Pages/OpenCityHall.aspx?pd_url=http%3a//www.peakdemocracy.com/portals/116/forum_home?admin_mode%3Doff#peak_democracy) with the extension I can see one root level annotation and two page comments (indicated with the icon next to the user name)

Page comments are annotations created for a uri, they have no target field (only an uri) and currently they're not counted into any buckets/tabs. [Example page comment](https://api.hypothes.is/annotations/-gQ5QnjiTP2bi20O4rsotQ)

That is why there is only one tab/bucket with the number 1 on it despite the sidebar shows three annotations. There was some brainstorming about page comments here: #1984 and here #1386 This is the answer for (4)

About (2) and (3) namely our search api, we have escaping problems here. I think actually we're double escaping some part of the uri.

So if I use this:
https://hypothes.is/api/search?uri=http://www.a2gov.org/services/Pages/OpenCityHall.aspx?pd_url=http%3a//www.peakdemocracy.com/portals/116/forum_home?admin_mode%3Doff#peak_democracy

It does give back the three annotations.
However the stream escapes with uri further and 
https://hypothes.is/stream?q=uri:pd_url%3Dhttp%253a%2F%2Fwww.peakdemocracy.com%2Fportals%2F116%2Fforum_home%3Fadmin_mode%253Doff%23peak_democracy#peak_democracy
it is no longer working.

I'll open a separate issue for this and let this issue handle the anchoring problems.

**Comment by gergely-ujvari** [comment]
As for the [arstechnica](https://via.hypothes.is/h/http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/) page, I cannot see any annotations on either links.

**Comment by judell** [comment]
I just made an annotation to http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/.

https://hypothes.is/stream?q=uri:http:%2F%2Farstechnica.com%2Fgaming%2F2015%2F04%2Fvalve-software-co-founder-gabe-newell-becomes-a-video-game-character%2F -> 1 annotation (mine)

https://hypothes.is/api/search?uri=http%3A%2F%2Farstechnica.com%2Fgaming%2F2015%2F04%2Fvalve-software-co-founder-gabe-newell-becomes-a-video-game-character%2F -> 1 annotation (mine)

https://via.hypothes.is/h/http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/ -> 1 annotation (mine)

Everyone see the same?

@dwhly can you repeat your experiment?

**Comment by dwhly** [comment]
Just ran saucelabs w/ safari/MacOS.  I see two annotations indicated in the sidebar.  One from Jakehart and the other from Buddha.

**Comment by gergely-ujvari** [comment]
I see the same as @judell

**Comment by judell** [comment]
Let's nail this. To be clear, we are all talking about this URL, right?

http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/.

I repeated the above using a stream query, and api query, and the extension. Results in this gist: https://gist.github.com/judell/341a98f3a73296bd5e54

@gergely-ujvari you are seeing exactly the same for that URL?

@dwhly you are seeing completely different results for that URL? Can you put them into a gist like mine for comparison?

My hunch is that this isn't only a Safari thing, I've been reporting inconsistencies like this for days, initially on GitHub but then I took the reporting to screencasts linked from the internal spreadsheet because I don't think we yet know even how to characterize what issue(s) are going on here, and we need to gather and collate more evidence before we can write a proper GH issue.

**Comment by tilgovi** [comment]
@judell from your gist I can't tell what, if anything, is wrong. It shows three different views of the same thing.

**Comment by judell** [comment]
The question is will @dwhly's version of that gist agree or disagree with
mine.

**Comment by judell** [comment]
Maryann's report confirms: In Chrome she sees what other Chrome users see, in Safari what other Safari users see, but Chrome and Safari appear to see two different sets of annotations.

--------------Maryann report--------

Well that's odd.  The first one is:

JakeHart6 days ago
As reported at Eurogamer, designer Barney Cumming took to the game's development blog on Thursday to post a brief yet tantalizing GIF of his latest in-game creation, and it appeared to portray Newell as a playable "boss" in the game. Crawl, which is currently available on Steam's Early Access service, works by having four players battle with each other to take command of the game's "hero" role—meaning everyone who's not the human must invade the bodies of evil creatures to make the hero's life miserable.
Annotation test... will soon be deleted.

and the second one is by Buddha

But I got to it by clicking on an annotation made by you that is not the same snippet.

Buddha6 days ago
The only other major, digital version of Newell we can recall ever appearing in a game came in the form of a fan-made Dota 2 mod of the game's shopkeeper, but that mod is no longer available through the Steam Workshop.
Another test!

I don't see yours.

Actually, it's not odd at all.  It's the same problem I've been seeing.  I'm assuming you did not make your annotation in Safari?  Because I see it in Chrome.  I don't see the other 2 in Chrome.  I don't think annotations transfer.

**Comment by gergely-ujvari** [comment]
I think we have a search api problem here with via/not via links. I checked the [Gabe Newell link](https://via.hypothes.is/h/http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/) with Chrome (both via via and with extension) and they only find one result here [d1NBzR6JQTmWwvfuxvT2uw](https://hypothes.is/api/annotations/d1NBzR6JQTmWwvfuxvT2uw) - made by @judell

But I see that there is [one more annotation](https://hypothes.is/api/annotations/Seebh46YT6ulQAMVuF3kGw) on this page, made by Buddha
and [one more](https://hypothes.is/api/annotations/r8CR5Ql0TPGMJZ_THE3FsQ) made by @RawKStar77 

The difference is that these two were made with via and it seems our search api fails to get them.
https://hypothes.is/api/search?limit=200&offset=0&order=asc&sort=created&uri=https:%2F%2Fvia.hypothes.is%2Fh%2Fhttp:%2F%2Farstechnica.com%2Fgaming%2F2015%2F04%2Fvalve-software-co-founder-gabe-newell-becomes-a-video-game-character%2F -> zero results
https://hypothes.is/api/search?uri=https://via.hypothes.is/h/http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/# -> zero results
https://hypothes.is/api/search?uri=http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/# -> 1 result

**Comment by judell** [comment]
This looks promising. I don't think it's the whole story, it feels like
there are a few different things going on, but eliminating this will surely
help.

**Comment by judell** [comment]
An example where the via/not-via distinction seems orthogonal (i.e. no via urls in the api view)

https://twitter.com/KenNeumeister/status/582190731592396800 

https://hypothes.is/api/search?uri=https%3A%2F%2Ftwitter.com%2FKenNeumeister%2Fstatus%2F582190731592396800 -> 2

https://hypothes.is/stream?q=uri:https:%2F%2Ftwitter.com%2FKenNeumeister%2Fstatus%2F582190731592396800 -> 2

https://via.hypothes.is/h/https://hypothes.is/stream?q=uri:https:%2F%2Ftwitter.com%2FKenNeumeister%2Fstatus%2F582190731592396800 -> 2

https://via.hypothes.is/h/https://twitter.com/KenNeumeister/status/582190731592396800 -> redirects to...
https://via.hypothes.is/h/20150414191427/https://twitter.com/KenNeumeister/status/582190731592396800 -> 3!

h bookmarklet -> 1

extension -> 1

wtf?

**Comment by gergely-ujvari** [comment]
Some news from the debugging. I'll still use the [Gabe Newell link](/http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/) as an example.

The fundamental differences between the annotations created by [me](https://hypothes.is/api/annotations/WRmN6_VtTW6D-V-bcWSnhA) or [@judell](https://hypothes.is/api/annotations/d1NBzR6JQTmWwvfuxvT2uw) and between the annotations created by [buddha](https://hypothes.is/api/annotations/Seebh46YT6ulQAMVuF3kGw) and [@RawKStar77](https://hypothes.is/api/annotations/r8CR5Ql0TPGMJZ_THE3FsQ) is the uri.

The last two (and probably safari) was a hashmark character in the end of the uri (i.e. `http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/#`) while the annotations created by chrome does not have that. (i.e.: `http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/`)

Now, somewhere we shallow the hashmark, because if I make a search api query like this: `http://localhost:5000/api/search?uri=http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/#` 

It'll be converted to the following ES query:
`
  {"sort": [{"updated": {"order": "desc", "ignore_unmapped": 'true'}}],
  "query": {
     "bool": {
       "must": [{
         "match": {
           "uri": "http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/"
   }}]}}, "from": 0, "size": 20}
`

So we shallow the hashmark character for search but we save the character into the uri of the annotation. That is still not the whole picture but getting closer, because running the ES query with the hashmark gives back the missing (from chrome) annotations.

`
`
  {"sort": [{"updated": {"order": "desc", "ignore_unmapped": 'true'}}],
  "query": {
     "bool": {
       "must": [{
         "match": {
           "uri": "http://arstechnica.com/gaming/2015/04/valve-software-co-founder-gabe-newell-becomes-a-video-game-character/#"
   }}]}}, "from": 0, "size": 20}
`

**Comment by gergely-ujvari** [comment]
From IRC:
@dwhly found an another example for this. [The annotation:Q_BrscxvSbqIeR_3lMMXAg]( https://hypothes.is/api/annotations/Q_BrscxvSbqIeR_3lMMXAg) doesn't appear in the [page](https://via.hypothes.is/h/http://boundary2.org/2015/04/15/a-darked-warped-reflection/)

The uri of the annotations has hashmark in the end: `http://boundary2.org/2015/04/15/a-darked-warped-reflection/#`

**Comment by judell** [comment]
@gergely-ujvari Looking more hopeful, thanks and please keep us posted!

**Comment by gergely-ujvari** [comment]
I think we can do a quick workaround by adding a line to the `uri_index` filter, a new pattern to capture:
`'(.+)\\#'` -> this removes the sole hashmarks from the end at indexing. Locally this is working.
@tilgovi: What do you think? Let's make a reality check here.

**Comment by tilgovi** [comment]
Seems sensible.

**Comment by tilgovi** [comment]
After looking at #2169 and thinking that it shouldn't be the solution I dug into our server and discovered that it seems like I never re-indexed once a7229ebf598807a771a50c365df9a8497a3810b6 went live.

Having done so now, I think this might be fixed.

**Comment by tilgovi** [comment]
I'm tentatively going to close this and leave #2076 open for Safari issues. Going through the comments above and trying to reproduce I'm not seeing the problem here anymore.

**Comment by judell** [comment]
Crossing fingers and touching wood. Thx.

**Comment by gergely-ujvari** [comment]
> After looking at #2169 <https://github.com/hypothesis/h/pull/2169> and
> thinking that it shouldn't be the solution I dug into our server and
> discovered that it seems like I never re-indexed once a7229eb
> <https://github.com/hypothesis/h/commit/a7229ebf598807a771a50c365df9a8497a3810b6>
> went live.
>
Ooh, now that is a problem.
>
> Having done so now, I think this might be fixed.
>
Yeah, this might solve many things here.
